dalm/training/generator_only/trainer.py

```diff
@@ -0,0 +1,274 @@
+    gradient_accumulation_steps,
+    gradient_checkpointing,
+    group_by_length,
+    packing,
+    lora_alpha,
+    lora_dropout,
+    lora_r,
+    learning_rate,
+    lr_scheduler_type,
+    num_warmup_steps,
+    weight_decay,
+    optimizer_type,
+    output_dir,
+    log_freq,
+    neftune_noise_alpha=5,
+    log_with="wandb",
+):
+    tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
+    tokenizer.pad_token = tokenizer.eos_token
+    tokenizer.padding_side = "right"  # Fix weird overflow issue with fp16 training
+
+    bnb_config = BitsAndBytesConfig(load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=torch.bfloat16)
+
+    base_model = AutoModelForCausalLM.from_pretrained(
+        model_name,
+        quantization_config=bnb_config,
+        device_map={"": Accelerator().local_process_index},
+        trust_remote_code=True,
+        use_auth_token=True,
+    )
+
+    base_model.config.use_cache = False
+
+    peft_config = LoraConfig(
+        r=lora_r,
+        lora_alpha=lora_alpha,
+        lora_dropout=lora_dropout,
+        target_modules=["q_proj", "v_proj"],
+        bias="none",
+        task_type="CAUSAL_LM",
+    )
+
+    training_args = TrainingArguments(
+        output_dir=output_dir,
+        per_device_train_batch_size=per_device_train_batch_size,
+        gradient_accumulation_steps=gradient_accumulation_steps,
+        per_device_eval_batch_size=per_device_eval_batch_size,
+        learning_rate=learning_rate,
+        logging_steps=logging_steps,
+        num_train_epochs=num_train_epochs,
+        report_to=log_with,
+        save_strategy="epoch",
+        evaluation_strategy="steps",
+        eval_steps=eval_steps,
+        group_by_length=group_by_length,
+        lr_scheduler_type=lr_scheduler_type,
+        warmup_steps=num_warmup_steps,
+        optim=optimizer_type,
+        bf16=True,
+        remove_unused_columns=False,
+        run_name="generator_tuning",
+    )
+
+    def prepare_sample_text(example):
+        """Prepare the text from a sample of the dataset."""
+        text = tokenizer.apply_chat_template(example["messages"], tokenize=False)
+        return text
+
+    train_dataset, eval_dataset = create_datasets(
+        dataset_name,
+        split,
+        size_valid_set,
+        streaming,
+        shuffle_buffer,
+        seq_length,
+        num_workers,
+        tokenizer,
+        prepare_sample_text,
+    )
+
+    trainer = SFTTrainer(
+        model=base_model,
+        train_dataset=train_dataset,
+        eval_dataset=eval_dataset,
+        peft_config=peft_config,
+        packing=packing,
+        max_seq_length=seq_length,
+        tokenizer=tokenizer,
+        args=training_args,
+        neftune_noise_alpha=neftune_noise_alpha,
+    )
+
+    trainer.train()
+    trainer.save_model(output_dir)
+
+    output_dir = os.path.join(output_dir, "final_checkpoint")
+    trainer.model.save_pretrained(output_dir)
+
```

Comment:
It might be good to add a logging statement statement tell the user where their final model was saved since the output directory has a default value, and they may not know what that value is.